dali/kernels/imgproc/convolution/filter_gpu.cuh

```diff
@@ -0,0 +1,1094 @@
+  using StaticConfigT = filter::StaticConfig<axes>;
+  using BlockSetupFactoryT = filter::AdaptiveBlock<StaticConfigT>;
+  using BlockSetupT = typename BlockSetupFactoryT::BlockSetup;
+  using StaticBlockFactoryT = filter::StaticBlock<StaticConfigT>;
+  using StaticBlockT = typename StaticBlockFactoryT::BlockSetup;
+  using GridSetupT = filter::GridSetup<StaticConfigT>;
+  using SampleDescT = filter::SampleDesc<Out, In, W, Intermediate, axes>;
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim>& out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageGPU, const W, axes>& filters,
+           const span<const ivec<axes>> anchors, boundary::BoundaryType border_type,
+           const TensorListView<StorageGPU, const In, 0>& fill_values = {}) {
+    auto num_samples = in.shape.num_samples();
+    assert(out.num_samples() == num_samples && filters.num_samples() == num_samples &&
+           anchors.size() == num_samples);
+    assert(fill_values.num_samples() == num_samples || fill_values.num_samples() == 0);
+
+    int max_total_workspace;
+    SampleDescT* samples_desc_dev;
+    SetupSampleDescs(max_total_workspace, out, in, filters, anchors);
+    std::tie(samples_desc_dev) = ctx.scratchpad->ToContiguousGPU(ctx.gpu.stream, samples_desc_);
+    auto grid_setup = PrepareGridSetup(out, in);
+    RunKernel(ctx, border_type, fill_values, [&](auto&& block_setup) {
+      return [&](auto&& loader) {
+        filter::filter<<<grid_setup.kernel_setup(), StaticConfigT::threadblock_size,
+                         max_total_workspace, ctx.gpu.stream>>>(samples_desc_dev, block_setup,
+                                                                loader, grid_setup);
+        CUDA_CALL(cudaGetLastError());
+      };
+    });
+  }
+
+ protected:
+  template <typename T>
+  vec<axes, T> ShapeAsVec(const TensorShape<ndim>& shape) {
+    return shape2vec<axes, T>(skip_dim<sequence_dim>(skip_dim<channels_dim>(shape)));
+  }
+
+  template <typename Inshapes, typename FilterShapes>
+  void ValidateNumericLimits(const Inshapes& in_shapes, const FilterShapes& filter_shapes) {
+    const auto max_grid_logical_extents = StaticConfigT::max_grid_extents() * StaticConfigT::lanes;
+    // so that we can safely use grid extents as a stride in a for loop
+    const auto max_sample_extents = std::numeric_limits<int>::max() - max_grid_logical_extents + 1;
+    for (int sample_idx = 0; sample_idx < in_shapes.num_samples(); sample_idx++) {
+      const auto& in_shape = in_shapes[sample_idx];
+      const auto& filter_shape = filter_shapes[sample_idx];
+      int64_t num_frames = has_sequence_dim ? in_shape[sequence_dim] : 1;
+      int64_t num_channels = has_channel_dim ? in_shape[channels_dim] : 1;
+      const auto channels = cat(i64vec<1>{num_channels}, i64vec<axes - 1>{1});
+      auto in_extents = ShapeAsVec<int64_t>(in_shape) * channels;
```

Comment:
This looks quite cryptic... I think the following is easier to decipher than the variant 1-padded vector.
```suggestion
      auto in_extents = ShapeAsVec<int64_t>(in_shape) * channels;
      in_extents[0] *= num_channels;
```